---
title: "Computing Assignment – Bayesian phylogenetics and epidemiology"
output:
  html_document: default
  pdf_document: default
---

```{r setup, message=FALSE, warning=FALSE, results='hide'}
# List of required packages
required_packages <- c(
  "rmarkdown", "tinytex", "tidyverse", "ggtree", 
  "ggridges", "treeio", "coda", "cols4all")

# Install missing packages by comparing against the currently installed packages
missing_packages <- required_packages[!(required_packages %in% installed.packages()[,"Package"])]
if(length(missing_packages)) {
  install.packages(missing_packages)
}
#loads all the packages by applying the library function to each
lapply(required_packages, library, character.only = TRUE)
```

Candidate Number: 1073601

**Question 1**

From the trees produced in the R Markdown file propose how many independent CHIKV introductions into Kassala sparked and sustained the 2018 outbreak, and their time of occurrence. How much earlier from the onset of the current outbreak is this estimated to be, and what does this imply? Use tree visualisations and the distribution of relevant parameter estimates to explain your answer.

```{r}
# Load the MCC tree
beast_file <- ("MCC_trees/MCC_CHIKV_IndianOcean.tree")
beast_tree <- treeio::read.beast(beast_file)

# Plots a tree showing the latest possible Kassala MRCA 
ggtree(beast_tree, aes(color = outbreak), mrsd="2018-10-16") + theme_tree2() +
  scale_x_continuous(breaks = c(2002, 2004, 2006, 2008, 2010, 2012, 2014, 2016, 2018),
                     minor_breaks = seq(2004, 2019, 1)) +
  geom_point2(aes(subset = (node == 255)), shape = 21, size = 3, fill = 'darkblue') +
  scale_color_discrete_c4a_cat("friendly7") +
  theme(panel.grid.major = element_line(color = "black", linewidth = .1),
        panel.grid.minor = element_line(color = "grey", linewidth = .1),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank())
#Scale bars are added, and the outbreaks are represented using colour-blind friendly visuals
```

As you can see from the above graph, all the Kassala nodes are monophyletic and branch from an individual MRCA that is inferred to have been located within Kassala. This suggests that only one independent CHIKV introduction caused the 2018 outbreak, most likely an long term carrier, possibly asymptomatic, that had traveled to Sudan. The estimated MRCA sets a limit for the most recent point at which the Kassala cases could have begun spreading, but there may have been earlier transmission events that attached to the tree at the long basal branch that were not sampled. Taking this into account, to estimate a more accurate date at which the Kassala outbreak began, we can use the posterior distribution of the "Age of Kassala Outbreak MRCA" parameter.

```{r}
#Creating a data frame of the parameters with the burn-in removed
beast_log <- read.csv("MCMC_log_files/CHIKV_IndianOcean_BEAST_output_log.csv")
beast_log_burnin <- beast_log[floor(nrow(beast_log) * 0.1):nrow(beast_log),]


#Generating a 95% HPD of the Kassala MRCA
hpd <- paste0(round(HPDinterval(as.mcmc(beast_log_burnin$age.Kassala_outbreak.))[1], 2),
              ", ",
              round(HPDinterval(as.mcmc(beast_log_burnin$age.Kassala_outbreak.))[2], 2))

beast_log_burnin |>
  select(age.Kassala_outbreak.) |>
  pivot_longer(cols = c("age.Kassala_outbreak."), names_to = "clade") |>
  ggplot(aes(x = value, y = clade, fill = factor(after_stat(quantile)))) +
  stat_density_ridges(geom = "density_ridges_gradient",
                      calc_ecdf = TRUE,
                      quantiles = c(0.025, 0.975),
                      alpha = 0.2) +
  scale_fill_manual(name = "Probability",
                    values = c("#882255", "#88CCEE", "#882255"),
                    labels = c("Lower 2.5% PD", "95% HPD", "Upper 2.5% PD")) +
  scale_y_discrete(expand = expansion(add = c(.1, 2.5))) +
  annotate("text", x = 2017, y = 2, label = hpd) +
  theme_classic() + theme(axis.title.y = element_blank(),
                          axis.text.y = element_blank(),
                          axis.ticks.y = element_blank())
```

From this 95% HPD we can see that the estimated date of the Kassala MRCA is somewhere between April 2016 and May 2017, with January 2017 being the most probable point at which CHIKV began circulating (given the current set of sampled viral sequences). Of course, this is almost a year before the MRCA estimate from the tree above (around the start of 2018), suggesting that there were in fact transmission events occurring earlier that weren't sampled.

**Question 2**

Provide a description of the circulation of the entire CHIKV Indian Ocean-specific lineage from this exercise from the early 2000s to 2018. Include viral movements across countries and regions, and approximate dates when these viral movements occurred.

```{r}
# Spread across regions 
ggtree(beast_tree, aes(color = region), mrsd = "2018-10-16") + theme_tree2() +
  scale_x_continuous(
    breaks = c(2002, 2004, 2006, 2008, 2010, 2012, 2014, 2016, 2018),
    minor_breaks = seq(2004, 2019, 1)
  ) +
  scale_color_discrete_c4a_cat("friendly7") +
  theme(
    panel.grid.major = element_line(color = "black", size = .1),
    panel.grid.minor = element_line(color = "grey", size = .1),
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank()
  )
```

The above graph shows the spread of the Indian Ocean CHIKV lineage across regions. Using this, and the below graph showing spread to individual countries, we can build a picture of the spread of the virus. As the name implies, this lineage originally started spreading as an epidemic around 2005 in the islands of La Reunion and Comoros, and was a sub-lineage of the earlier Asian CHIKV genotype. The original transmission origin for this outbreak is hard to infer, but from the below graph we can see it must have been somewhere between these islands and mainland Africa in Kenya. The increased severity of this outbreak has been attributed to a mutation that allowed the virus to acquire additional mosquito species as vectors and expand beyond its native range (Schuffenecker et al, 2006).

The virus then continued to spread around the Indian ocean islands, including to Mauritius, Mayotte and the Seychelles around late 2006. While this epidemic subsided, an outbreak began spreading in India around 2007/8, spreading to neighboring Sri Lanka towards the end of 2008 and even into South East Asia with cases reported in Singapore around 2009. The virus continually subsided and re-emerged within India from 2009 till as late as 2016, spreading to Bangladesh in 2012 and Yemen in 2011, and this Indian re-emergence eventually led to cases in Kenya in 2017. Simultaneously, another Indian strain led to outbreaks in Pakistan in 2016. The origin of the Sudan 2018 epidemic is also thought to have originated from an Indian infection, spreading to Kassala in early 2018.

```{r}
ggtree(beast_tree, aes(color = country), mrsd="2018-10-16") + theme_tree2() +
  scale_x_continuous(breaks = c(2002, 2004, 2006, 2008, 2010, 2012, 2014, 2016, 2018),
                     minor_breaks = seq(2004, 2019, 1)) +
  scale_color_discrete_c4a_cat("classic20") +
  theme(panel.grid.major = element_line(color = "black", size = .1),
        panel.grid.minor = element_line(color = "grey", size = .1),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank())
```

References for this question:

Schuffenecker, I., Iteman, I., Michault, A., Murri, S., Frangeul, L., Vaney, M.-C., Lavenir, R., Pardigon, N., Reynes, J.-M., Pettinelli, F., Biscornet, L., Diancourt, L., Michel, S., Duquerroy, S., Guigon, G., Frenkiel, M.-P., Bréhin, A.-C., Cubito, N., Desprès, P. and Kunst, F. (2006). Genome Microevolution of Chikungunya Viruses Causing the Indian Ocean Outbreak. PLoS Medicine, 3(7), p.e263. <doi:https://doi.org/10.1371/journal.pmed.0030263>.

**Question 3**

Imagine a fictitious scenario where, following the 2018 outbreak, CHIKV infections continued to be reported in Kassala throughout 2019. We have expanded the original data set to include 7 new CHIKV genome sequences (please note that, since this scenario is fictitious, the 2019 Kassala sequences are synthetic and created for the purpose of this exercise). A similar analysis to the 2018 outbreak was performed, the resulting tree can be found in the MCC_CHIKV_IndianOcean_updated.tree file and the log file can be found as CHIKV_IndianOcean_updated_BEAST_output_log.csv. Are the 2019 cases a continuation of the 2018 epidemic or a different outbreak? Justify your answer with a visualisation of the tree and its interpretation.

```{r, warning=F}
# Loading the updated tree
updated_beast_file <- ("MCC_trees/MCC_CHIKV_IndianOcean_updated.tree")
updated_beast_tree <- treeio::read.beast(updated_beast_file)

# Plots a tree showing the latest possible Kassala MRCA (note later mrsd)
ggtree(updated_beast_tree, aes(color = outbreak), mrsd="2019-10-16") + theme_tree2() +
  scale_x_continuous(breaks = c(2002, 2004, 2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020),
                     minor_breaks = seq(2003, 2019, 1)) +
  scale_color_discrete_c4a_cat("friendly7") +
  theme(panel.grid.major = element_line(color = "black", linewidth = .1),
        panel.grid.minor = element_line(color = "grey", linewidth = .1),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank())
#Scale bars are updated to accommodate the more recent transmissions, and the outbreaks are represented using colour-blind friendly visuals
```

As the updated tree shows, the 2019 cases are a mix of both a continuation of the 2018 Kassala lineage and also a separate new introduction. This can be seen by the fact that one sampled case branches out of the 2018 Kassala lineage, implying that it was a continuation. The remaining cases, however, form a sister clade to the 2018 outbreak, suggesting that this was a distinct introduction from India to Sudan.

**Question 4**

The updated log file contains parameter estimates for the MRCAs for the 2018 (age(Kassala_2018)) and 2019 (age(Kassala_2019)) sequences. From a comparison of these dates and the specific times when cases started to increase for each season (July in 2018 and April in 2019), is there evidence for new CHIKV importations happening during the same time every year? Why? Are importation times related to the timing when epidemics start to take off?

```{r}
#Creating an updated data frame of the new parameters with the burn-in removed
updated_beast_log <- read.csv("MCMC_log_files/CHIKV_IndianOcean_updated_BEAST_output_log.csv")
updated_beast_log_burnin <- updated_beast_log[floor(nrow(updated_beast_log) * 0.1):nrow(updated_beast_log),]
```

```{r}
#Creating a violin plot to compare the MRCA of the 2018 and 2019 outbreak
updated_beast_log_burnin |>
  select(age.Kassala_2019., age.Kassala_2018.) |>
  pivot_longer(cols = c("age.Kassala_2019.", "age.Kassala_2018."), names_to = "clade_age") |>
  ggplot() + geom_violin(aes(x = value, y = clade_age)) +
  theme_classic()
```

From the above violin plot, we can see that the MRCA for the 2019 outbreak is estimated to be between mid to late 2018, while the 2018 outbreak's MRCA is around early to mid 2017. It is interesting to note that the additional 2019 samples have changed the 2018 outbreak MRCA, presumably shifting this date forward in response to the later sample that branched out of the 2018 cases. This indicates that the novel CHIKV importations do not occur at the same point each year - 2018 outbreak introduction occurred during the spring (of 2017) whereas the 2019 outbreak introduction was in the autumn of 2018.

Comparing the estimated MRCAs of the two outbreaks with the point at which the cases start proliferating exponentially in the population, there is a clear difference: the 2019 outbreak took around 6-8 months to reach this stage, whereas the 2018 outbreak took more than a year. This is because the timing of the CHIKV introduction corresponds to the rate of its proliferation. For example, the time between the 2019 MRCA and outbreak is so short because it was introduced in the wet autumn, with significantly more mosquitos to increase spread. Conversely, the 2018 MRCA introduction in the dry spring would be spread slower, explaining the increased lag time.

**Question 5**

Explore the imported BEAST tree (beast_tree) object and extract the counts of node
locations by country. Through which country or countries has this CHIKV lineage
circulated more frequently? Have these trends changed over time? Use attributes of
the tree object within R such as ‘label’, ‘isTip’, ‘country’ and ‘height’ to answer these
questions. Consider that plots like histograms and time series can be useful to explore
these questions.


```{r}
beast_tree@data

beast_tree_data <- fortify(beast_tree)


# Count node locations by country
location_counts <- beast_tree_data %>%
  filter(!is.na(country)) %>%  # Remove NA values
  count(country, name = "count") %>%
  arrange(desc(count))

ggplot(location_counts, aes(x = reorder(country, -count), y = count, fill = country)) +
  geom_bar(stat = "identity") +
  theme_classic() +
  labs(x = "Country", y = "Node Count", title = "Node Counts by Country") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

